common/src/KokkosKernels_Error.hpp

```diff
@@ -53,4 +53,71 @@ inline void hip_internal_safe_call(hipError_t e, const char *name,
 }  // namespace Impl
 }  // namespace KokkosKernels
 
+/*
+ * Asserts and error checking macros/functions.
+ *
+ * KK_KERNEL** are for error checking within kokkos kernels.
+ *
+ * Any check with "assert" in the name is disabled for release builds
+ *
+ * For _MSG checks, the msg argument can contain '<<' if not a kernel check.
+ *
+ * This code is adapted from EKAT/src/ekat/ekat_assert.hpp
+ */
+
+// Internal do not call directly
+#define IMPL_THROW(condition, msg, exception_type)                      \
+  do {                                                                  \
+    if (!(condition)) {                                                 \
+      std::stringstream _ss_;                                           \
+      _ss_ << __FILE__ << ":" << __LINE__ << ": FAIL:\n" << #condition; \
+      _ss_ << "\n" << msg;                                              \
+      throw exception_type(_ss_.str());                                 \
+    }                                                                   \
+  } while (0)
+
+// SYCL cannot printf like the other backends quite yet
+#ifdef __SYCL_DEVICE_ONLY__
+#define IMPL_KERNEL_THROW(condition, msg)                               \
+  do {                                                                  \
+    if (!(condition)) {                                                 \
+      const __attribute__((opencl_constant)) char format[] =            \
+          "KERNEL CHECK FAILED:\n   %s %s\n";                           \
+      sycl::ext::oneapi::experimental::printf(format, #condition, msg); \
+      Kokkos::abort("");                                                \
+    }                                                                   \
+  } while (0)
+#else
+#define IMPL_KERNEL_THROW(condition, msg)                              \
+  do {                                                                 \
+    if (!(condition)) {                                                \
+      printf("KERNEL CHECK FAILED:\n   %s\n   %s\n", #condition, msg); \
+      Kokkos::abort("");                                               \
+    }                                                                  \
+  } while (0)
+#endif
+
+#ifndef NDEBUG
```

Comment:
I was not able to find any pre-existing assert/require macros that worked inside kernels, so I brought some in from another project that have worked well for us.